### PR TITLE
Fea ext lagrangian MeshGraphNet

### PR DESCRIPTION
**Pull request by zongyi-li**
# Modulus Pull Request for Lagrangian MeshGraphNet

## Description
We implemented Meshgraphnet for particle-based simulation on the water dataset based on
<https://github.com/google-deepmind/deepmind-research/tree/master/learning_to_simulate> in PyTorch.
It demonstrates how to train a Graph Neural Network (GNN) for evaluation of the
Lagrangian fluid.

In this project, we provide an example of Lagrangian mesh simulation for fluids. The Lagrangian mesh is particle-based, where vertices represent fluid particles and edges represent their interactions. Compared to an Eulerian mesh, where the mesh grid is fixed, a Lagrangian mesh is more flexible since it does not require tessellating the domain or aligning with boundaries. 

As a result, Lagrangian meshes are well-suited for representing complex geometries and free-boundary problems, such as water splashes and object collisions. However, a drawback of the Lagrangian mesh is that it typically requires smaller time steps to maintain physically valid simulations.

The main code consists of dataloader, train, and inference
- `modulus/datapipes/gnn/lagrangian_dataset.py`
- `examples/cfd/lagrangian_mgn/train.py`
- `examples/cfd/lagrangian_mgn/inference.py`

lagrangian_meshgraphnet

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [ ] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

## Dependencies

- tensorflow to load the datasets

### CODE REVIEW

**Comment by michaltakac** [comment]
This is really great! Was thinking that it would make sense to also have an example for lagrangian GNN based on DeepMind's paper... Great work!

File: modulus/datapipes/gnn/lagrangian_dataset.py [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        self.length = num_samples * (num_steps - 1)
+
+        path_metadata = os.path.join(data_dir, "metadata.json")
+        with open(path_metadata, "r") as file:
+            metadata = json.load(file)
+            self.dt = metadata["dt"]
+            self.radius = metadata["default_connectivity_radius"]
+            self.bound = metadata["bounds"][0]
+            self.dim = metadata["dim"]
+            self.vel_mean = torch.tensor(metadata["vel_mean"]).reshape(1, 2)
```
Comment (michaltakac) [inline modulus/datapipes/gnn/lagrangian_dataset.py:148]: @zongyi-li Use reshape(1, self.dim) to make it work with both 2D and 3D datasets.

File: examples/cfd/lagrangian_mgn/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# MeshGraphNet with Lagrangian mesh
+
+This example of Meshgraphnet for particle-based simulation on the water dataset based on
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/README.md:3]: "This is an example of ..."

File: modulus/datapipes/gnn/lagrangian_dataset.py [hunk 3]
```diff
@@ -0,0 +1,346 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
```
Comment (mnabian) [inline modulus/datapipes/gnn/lagrangian_dataset.py:1]: Please add unit tests for `lagrangian_dataset.py` and `lagrangian_reading_utils.py`.

File: examples/cfd/lagrangian_mgn/train.py [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+        self.dist = DistributedManager()
+        self.amp = cfg.amp
+        self.radius = cfg.radius
+        self.dt = cfg.dt
+        self.dim = cfg.dim
+        self.gravity = torch.zeros(self.dim, device=self.dist.device)
+        self.gravity[-1] = -9.8
+
+        # MGN with recompute_activation currently supports only SiLU activation function.
+        mlp_act = "silu"
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/train.py:54]: Why is this hard-coded?

File: examples/cfd/lagrangian_mgn/train.py [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+        if cfg.recompute_activation:
+            rank_zero_logger.info(
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/train.py:56]: I suggest raising an error instead

File: examples/cfd/lagrangian_mgn/train.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+        # backward pass
+        if self.amp:
+            self.scaler.scale(loss).backward()
+            self.scaler.step(self.optimizer)
+            self.scaler.update()
+        else:
+            loss.backward()
+            self.optimizer.step()
+
+    def l2loss(self, input, target, p=2, eps=1e-5):
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/train.py:204]: Is this needed? Since you are using MSE

File: examples/cfd/lagrangian_mgn/train.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+    # initialize distributed manager
+    DistributedManager.initialize()
+    dist = DistributedManager()
+
+    # Initialize loggers.
+    wandb.login(key=cfg.wandb_key)
+    initialize_wandb(
+        # project="meshgraphnet water-3D",
+        project="meshgraphnet",
+        entity="zongyi-li",
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/train.py:225]: Please make this configurable and remove your username

File: examples/cfd/lagrangian_mgn/train.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+            f"epoch: {epoch}, loss: {mean_loss:10.3e}, "
+            f"position loss: {mean_loss_pos:10.3e}, "
+            f"velocity loss: {mean_loss_vel:10.3e}, "
+            f"acceleration loss: {mean_loss_acc:10.3e}, "
+            f"time per epoch: {(time.time()-start):10.3e}"
+        )
+        wandb.log({"loss": mean_loss})
+        wandb.log({"loss_pos": mean_loss_pos})
+        wandb.log({"loss_vel": mean_loss_vel})
+        wandb.log({"loss_acc": mean_loss_acc})
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/train.py:261]: These can be combined into logging a single dictionary

File: examples/cfd/lagrangian_mgn/inference.py [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+            drop_last=False,
+        )
+
+        # instantiate the model
+        self.model = MeshGraphNet(
+            cfg.num_input_features,
+            cfg.num_edge_features,
+            cfg.num_output_features,
+            cfg.processor_size,
+            mlp_activation_fn="silu",
```
Comment (mnabian) [inline examples/cfd/lagrangian_mgn/inference.py:94]: This should be configurable

**Comment by zongyi-li** [comment]
Thanks for the comments! I have just addressed them and added a unit test `test/datapipes/test_lagrangian.py`. Please take a look!

File: modulus/datapipes/gnn/lagrangian_dataset.py (hunk 1 above)
Comment (zongyi-li) [reply modulus/datapipes/gnn/lagrangian_dataset.py]: Thanks! fixed

**Review by mnabian — APPROVED** [review]
LGTM
